spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -276,6 +270,30 @@ else if (e instanceof ProvisioningException) {
 		}
 	}
 
+	private void enhanceMessageChannel(MessageChannel inputChannel) {
+		((AbstractMessageChannel)inputChannel).addInterceptor(new ChannelInterceptorAdapter() {
+			@SuppressWarnings("unchecked")
+			@Override
+			public Message<?> preSend(Message<?> message, MessageChannel channel) {
+				if (message.getPayload() instanceof byte[]) {
+					if (!message.getHeaders().containsKey(BinderHeaders.NATIVE_HEADERS_PRESENT)
+							&& EmbeddedHeaderUtils.mayHaveEmbeddedHeaders((byte[]) message.getPayload())) {
+						MessageValues messageValues;
+						try {
+							messageValues = EmbeddedHeaderUtils.extractHeaders((Message<byte[]>) message, true);
+						}
+						catch (Exception e) {
+							logger.debug(EmbeddedHeaderUtils.decodeExceptionMessage(message),e);
```

Comment:
Why `debug` not `error`?
We have just caught an `Exception` though

Reply:
that's an old/existing code so while you may be right, it would needs to be fixed in another issue, another PR

Reply:
I changed it from `error` to `debug` since we are falling back and we don't know for sure that the headers are embedded - we only attempt to fall back if the first byte is 0xff - it is possible that a real message without embedded headers passes that condition (but the extract will fail).

Reply:
Before I added my fallback code, it was error because the binder expected embedded headers. This now allows the reception of messages from producers using any header mode (native, embedded or none).

Reply:
Thank you for explanation!

In SI I would request a comment for this kind of concerns. 😉 
Eventually we all may forget why it is debug on exception, so someone may change it unconditionally.

Reply:
Good point - https://github.com/spring-cloud/spring-cloud-stream/commit/b9884ae71a90bb65a5471d06ff1e716cf647c7e6